### DISCUSSION TITLE
Httpx includes ipv6 scope in Host header

### DISCUSSION THREAD

**Discussion by elupus** [encode/httpx#2180] [Potential Issue]
When requesting http paths with scope as part of the ipv6 address, this should stripped before being added to the http host header since it's fully local feature of the ipv6 stack.

Http over link local addresses is quite common on upnp based announces on link local networks.

Apache for example will reject the request: https://bz.apache.org/bugzilla/show_bug.cgi?id=35122
Lighthttpd: https://redmine.lighttpd.net/issues/2442

Example to reproduce:

```python
import httpx
import respx

@respx.mock
def test_example():
    my_route = respx.get("https://[fe80::8fd:8e9a:40d8:29a6%en0]/get")
    my_route.mock(return_value=httpx.Response(204))

    r = httpx.get(r'https://[fe80::8fd:8e9a:40d8:29a6%en0]/get')

    print(my_route.calls[0].request.headers["host"])
    assert my_route.calls[0].request.headers["host"] == "[fe80::8fd:8e9a:40d8:29a6]"

test_example()
```

---

- [ ] Initially raised as discussion #...

**Comment by elupus** [comment]
Ok. I must admit I'm a bit unsure of current RFC status regarding this. Se discussion here: https://bugs.python.org/issue23448

**Comment by lovelydinosaur** [comment]
It appears to me that we have the correct behaviour here...

Based on [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2) and other searching around, it appears that IPv6 addresses in the Host header should be formatted with the enclosing square brackets.

Which is what we do...

```python
import httpx


c = httpx.Client()
r = c.build_request("GET", "https://[fe80::8fd:8e9a:40d8:29a6]/get")
print(r.headers["host"])  # [fe80::8fd:8e9a:40d8:29a6]
```

Am I misunderstanding your issue here?

Is there a public address you can use to demonstrate the problem?

**Reply by elupus** [reply]
This is only a local scope issue. Routed IPs don't need this.

Notice the %en0 in the request. This is needed to make the request from the correct local network interface for link local ipv6. Windows machines may work without it, but POSIX won't.

However it's not something the target host should see. It's a rather stupid ipv6 choice for link local access.

Some info exist on Wikipedia for example: https://en.m.wikipedia.org/wiki/IPv6_address#zone_index
